### PR TITLE
Feature/out dacl

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Support for out direction in Dynamic acls.

# Impacts
Cisco Switches

# Delete branch after merge
YES

# Checklist
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

## Enhancements
* Support for out acls if the switch support it

### CODE REVIEW

File: lib/pf/Switch/Cisco/ASA.pm [hunk 1]
```diff
@@ -265,12 +274,17 @@ sub returnRadiusAdvanced {
             for ( my $loops = 0; $loops < $self->ACLsLimit; $loops++ ) {
                 last if (scalar @{$session->{'acl'}} == 1);
                 my $acl = shift @{$session->{'acl'}};
-                if ($acl !~ /^permit/i && $acl !~ /^deny/i) {
+                if ($acl !~ /^permit/i && $acl !~ /^deny/i && $acl !~ /^in\|/i && $acl !~ /^out\|/i) {
```
Comment (jrouzierinverse) [inline lib/pf/Switch/Cisco/ASA.pm]: The regex will not catch the following `in|bad`
change to 
```perl
$acl =~ /^((in|out)\|)?(permit|deny)/
```

File: lib/pf/Switch/Cisco/ASA.pm [hunk 2]
```diff
@@ -279,11 +293,16 @@ sub returnRadiusAdvanced {
         }
         if (scalar @{$session->{'acl'}} == 1) {
             my $acl = shift @{$session->{'acl'}};
-            if ($acl =~ /^permit/i || $acl =~ /^deny/i) {
-                push(@av_pairs, $self->returnAccessListAttribute($session->{'acl_num'})."=".$acl);
-                $logger->info("(".$self->{'_id'}.") Adding access list : $acl to the RADIUS reply");
-                $logger->info("(".$self->{'_id'}.") Added access lists to the RADIUS reply.");
-                $self->setRadiusSession($session);
+            if ($acl =~ /^permit/i || $acl =~ /^deny/i || $acl =~ /^in\|/i || $acl =~ /^out\|/i) {
```
Comment (jrouzierinverse) [inline lib/pf/Switch/Cisco/ASA.pm]: The regex will not catch the following `in|bad`
change to
```perl
$acl =~ /^((in|out)\|)?(permit|deny)/
```

File: lib/pf/Switch/Cisco/Catalyst_2960.pm [hunk 3]
```diff
@@ -511,12 +511,17 @@ sub returnRadiusAccessAccept {
                     my $acl_num = 101;
                     while($access_list =~ /([^\n]+)\n?/g){
                        my $acl = $1;
-                       if ($acl !~ /^permit/i && $acl !~ /^deny/i) {
+                       if ($acl !~ /^permit/i && $acl !~ /^deny/i && $acl !~ /^in\|/i && $acl !~ /^out\|/i) {
```
Comment (jrouzierinverse) [inline lib/pf/Switch/Cisco/Catalyst_2960.pm]: change to
```perl
$acl !~ /^((in|out)\|)?(permit|deny)/
```

File: lib/pf/Switch/Cisco/Catalyst_2960.pm [hunk 4]
```diff
@@ -585,12 +590,17 @@ sub returnRadiusAdvanced {
             for ( my $loops = 0; $loops < $self->ACLsLimit; $loops++ ) {
                 last if (scalar @{$session->{'acl'}} == 1);
                 my $acl = shift @{$session->{'acl'}};
-                if ($acl !~ /^permit/i && $acl !~ /^deny/i) {
+                if ($acl !~ /^permit/i && $acl !~ /^deny/i && $acl !~ /^in\|/i && $acl !~ /^out\|/i) {
```
Comment (jrouzierinverse) [inline lib/pf/Switch/Cisco/Catalyst_2960.pm]: change to
```perl
$acl !~ /^((in|out)\|)?(permit|deny)/
```

File: lib/pf/Switch/Cisco/Catalyst_2960.pm [hunk 5]
```diff
@@ -599,11 +609,16 @@ sub returnRadiusAdvanced {
         }
         if (scalar @{$session->{'acl'}} == 1) {
             my $acl = shift @{$session->{'acl'}};
-            if ($acl =~ /^permit/i || $acl =~ /^deny/i) {
-                push(@av_pairs, $self->returnAccessListAttribute($session->{'acl_num'})."=".$acl);
-                $logger->info("(".$self->{'_id'}.") Adding access list : $acl to the RADIUS reply");
-                $logger->info("(".$self->{'_id'}.") Added access lists to the RADIUS reply.");
-                $self->setRadiusSession($session);
+            if ($acl =~ /^permit/i || $acl =~ /^deny/i || $acl =~ /^in\|/i || $acl =~ /^out\|/i) {
```
Comment (jrouzierinverse) [inline lib/pf/Switch/Cisco/Catalyst_2960.pm]: change to
```perl
$acl =~ /^((in|out)\|)?(permit|deny)/
```

File: lib/pf/Switch/Dell/N1500.pm [hunk 6]
```diff
@@ -355,12 +355,17 @@ sub returnRadiusAccessAccept {
                     my $acl_num = 101;
                     while($access_list =~ /([^\n]+)\n?/g){
                         my $acl = $1;
-                        if ($acl !~ /^permit/i && $acl !~ /^deny/i) {
+                        if ($acl !~ /^permit/i && $acl !~ /^deny/i && $acl !~ /^in\|/i && $acl !~ /^out\|/i) {
```
Comment (jrouzierinverse) [inline lib/pf/Switch/Dell/N1500.pm]: change to
```perl
$acl !~ /^((in|out)\|)?(permit|deny)/
```

File: lib/pf/Switch/Dell/N1500.pm [hunk 7]
```diff
@@ -509,12 +514,17 @@ sub returnRadiusAdvanced {
             for ( my $loops = 0; $loops < $self->ACLsLimit; $loops++ ) {
                 last if (scalar @{$session->{'acl'}} == 1);
                 my $acl = shift @{$session->{'acl'}};
-                if ($acl !~ /^permit/i && $acl !~ /^deny/i) {
+                if ($acl !~ /^permit/i && $acl !~ /^deny/i && $acl !~ /^in\|/i && $acl !~ /^out\|/i) {
```
Comment (jrouzierinverse) [inline lib/pf/Switch/Dell/N1500.pm]: change to
```perl
$acl !~ /^((in|out)\|)?(permit|deny)/
```

File: lib/pf/Switch/Dell/N1500.pm [hunk 8]
```diff
@@ -524,12 +534,17 @@ sub returnRadiusAdvanced {
         }
         if (scalar @{$session->{'acl'}} == 1) {
             my $acl = shift @{$session->{'acl'}};
-            if ($acl =~ /^permit/i || $acl =~ /^deny/i) {
-                push(@av_pairs, $self->returnAccessListAttribute($session->{'acl_num'})."=".$acl);
-                $logger->info("(".$self->{'_id'}.") Adding access list : $acl to the RADIUS reply");
-                $logger->info("(".$self->{'_id'}.") Added access lists to the RADIUS reply.");
-                $self->setRadiusSession($session);
-                push(@av_pairs, "ACS:CiscoSecure-Defined-ACL=$mac-".$session_id);
+            if ($acl =~ /^permit/i || $acl =~ /^deny/i || $acl =~ /^in\|/i || $acl =~ /^out\|/i) {
```
Comment (jrouzierinverse) [inline lib/pf/Switch/Dell/N1500.pm]: change to
```perl
$acl =~ /^((in|out)\|)?(permit|deny)/
```

File: html/pfappserver/lib/pfappserver/Util/ACLs.pm [hunk 9]
```diff
@@ -20,6 +20,9 @@ sub _validate_acl {
     my ($field) = @_;
     my $acl = $field->value;
     if ($acl) {
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Util/ACLs.pm:22]: We should also test the acl limit of this switch, like if it's support acls/dlacls and it reach the limit then trigger an error.

File: html/pfappserver/lib/pfappserver/Form/Config/Switch.pm [hunk 10]
```diff
@@ -490,9 +490,41 @@ sub addRoleMapping {
...
+    my $type_field = $self->field('type');
+    my $type = $type_field->value;
+    if (!defined $type) {
+        return undef;
+    }
+
+    my $value = $self->value;
+    my $module = pf::SwitchFactory::getModule($type);
+    if ($module->require() ) {
+        return $module->new($value);
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Form/Config/Switch.pm:522]: This doesn't return the switch config , all the field are empty and it fail after in the code.

**Pull request merged**